c/tests/integration/driver_test.c

```diff
@@ -0,0 +1,437 @@
+
+#define FAILED() check_error_may_print(__FILE__, __LINE__)
+
+bool test_database_management() {
+    const char databaseName[] = "test_database_management";
+
+    Connection* connection = NULL;
+    DatabaseManager* databaseManager = NULL;
+
+    bool success = false;
+
+    connection = connection_open_core(TYPEDB_CORE_ADDRESS);
+    if (FAILED()) goto cleanup;
+
+    databaseManager = database_manager_new(connection);
+    if (FAILED()) goto cleanup;
+
+    delete_database_if_exists(databaseManager, databaseName);
+    if (FAILED()) goto cleanup;
+
+    databases_create(databaseManager, databaseName);
+    if (FAILED()) goto cleanup;
+
+    if (!databases_contains(databaseManager, databaseName)) {
+        fprintf(stderr, "databases_contains(\'%s\') failed\n", databaseName);
+        goto cleanup;
+    }
+
+    bool foundDB = false;
+    DatabaseIterator* it = databases_all(databaseManager);
+    Database* database = NULL;
+    while (NULL != (database = database_iterator_next(it))) {
+        char* name = database_get_name(database);
+        foundDB = foundDB || (0 == strcmp(databaseName, name));
+        free(name);
+        database_drop(database);
+    }
+    database_iterator_drop(it);
+
+    if (!foundDB) {
+        fprintf(stderr, "Did not find database \'%s\' in list of databases\n", databaseName);
+        goto cleanup;
+    }
+
+    success = true;
+cleanup:
+    delete_database_if_exists(databaseManager, databaseName);
+    check_error_may_print(__FILE__, __LINE__);
+    database_manager_drop(databaseManager);
+    connection_close(connection);
+    return success;
+}
+
+bool test_query_schema() {
+    const char databaseName[] = "test_query_schema";
+
+    Connection* connection = NULL;
+    DatabaseManager* databaseManager = NULL;
+    Session* session = NULL;
+    Transaction* transaction = NULL;
+    Options* opts = NULL;
+
+    bool success = false;
+
+    // Set up connection & database
+    connection = connection_open_core(TYPEDB_CORE_ADDRESS);
+    if (FAILED()) goto cleanup;
+
+    databaseManager = database_manager_new(connection);
+    if (FAILED()) goto cleanup;
+
+    delete_database_if_exists(databaseManager, databaseName);
+    if (FAILED()) goto cleanup;
+
+    databases_create(databaseManager, databaseName);
+    if (FAILED()) goto cleanup;
+
+    opts = options_new();
+    if (FAILED()) goto cleanup;
+
+    // test schema queries
+    {
+        session = session_new(databaseManager, databaseName, Schema, opts);
+        if (FAILED()) goto cleanup;
+
+        transaction = transaction_new(session, Write, opts);
+        if (FAILED()) goto cleanup;
+
+        query_define(transaction, "define name sub attribute, value string;", opts);
+        if (FAILED()) goto cleanup;
+
+        ConceptMapIterator* it = query_match(transaction, "match $t sub thing;", opts);
+        ConceptMap* conceptMap;
+        bool foundName = false;
+        while (NULL != (conceptMap = concept_map_iterator_next(it))) {
```

Comment:
this may have to check `has_next` and then use `next` separately, without relying on getting a NULL back - if so, please update all the places

Reply:
Good idea. I checked, but these don't have a `has_next`